### PR TITLE
Fix malformed asset_url in ActionController::Renderer

### PR DESCRIPTION
**Pull request by tzabaman**
### Summary
#28151
Fix malformed asset_url when rendering a template with ActionController::Renderer

### CODE REVIEW

File: actionpack/test/controller/renderer_test.rb [hunk 1]
```diff
@@ -96,13 +95,20 @@ class RendererTest < ActiveSupport::TestCase
...
     content = renderer.render inline: "<%= request.ssl? %>"
 
     assert_equal "true", content
   end
 
+  test "return valid asset url with defaults" do
+    renderer = ApplicationController.renderer
+    content  = renderer.render template: "test/render_file_with_asset"
+
+    assert_equal "http://example.org/asset.jpg", content
```
Comment (Erol) [inline actionpack/test/controller/renderer_test.rb]: Can you also test the scenario where a renderer is initialized with `url_scheme: "https"`?

Comment (tzabaman) [inline actionpack/test/controller/renderer_test.rb]: @Erol Done!

**Comment by pixeltrix** [comment]
What's the need to switch to `url_scheme` as the API? Has something changed elsewhere? I assume that `https: true` used to work? Did something change in Rack?

**Comment by tzabaman** [comment]
@pixeltrix Take a look at issue #28151 . The problem is described inside the conversation. I don't know if something has changed in Rack but the asset_url was not working properly. asset_url was correct only if https was set to true. We had a malformed url if https was set to false. So scheme_url with params 'http' or 'https' seemed to be solving the problem so https was was no longer necessary!

**Comment by pixeltrix** [comment]
@tzabaman that's a good first step - you've found out the how, but we need to know when and why too.

e7633bedf897bb24ce668ac9c5df6bf88a58ff7e114d27606a756f4c4888a3f1

So the first step is to write a standalone test case that we can use to test against multiple versions:

``` ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", "5.0.2"
end

require "rack/test"
require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = File.dirname(__FILE__)
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_token    = "secret_token"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
  end
end

class ApplicationController < ActionController::Base; end

require "minitest/autorun"

class BugTest < Minitest::Test
  def test_default_renderer_with_asset_url
    template = "<%= asset_url('avatar.jpg') %>"
    renderer = ApplicationController.renderer
    rendered = renderer.render inline: template

    assert_equal "http://example.org/avatar.jpg", rendered
  end
end
```

So starting with Rails 5.0.2 what do we get?

```
  1) Failure:
BugTest#test_default_renderer_with_asset_url [rails-28151.rb:41]:
--- expected
+++ actual
@@ -1 +1 @@
-"http://example.org/avatar.jpg"
+"http://://example.org:80/avatar.jpg"
```

Okay, so what about Rails 5.0.0? Change the gemspec to the following:

``` ruby
gemfile(true) do
  source "https://rubygems.org"
  gem "rails", "5.0.0"
end
```

Re-run the test and what do we get?

```
  1) Failure:
BugTest#test_default_renderer_with_asset_url [rails-28151.rb:41]:
--- expected
+++ actual
@@ -1 +1 @@
-"http://example.org/avatar.jpg"
+"http://://example.org:80/avatar.jpg"
```

Nope, still broken. Looking at the history for `actionpack/lib/action_controller/renderer.rb` we can see a significant refactoring took place in 2db7304 - what if we go back earlier than that and test it worked when first committed. To do this we need to use the following gemspec:

``` ruby
gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails", ref: "801e399" # commit where feature was introduced
  gem "arel", github: "rails/arel", ref: "56040f2" # arel master HEAD at the time
end
```

So surely this will work, right?

```
  1) Failure:
BugTest#test_default_renderer_with_asset_url [rails-28151.rb:41]:
--- expected
+++ actual
@@ -1 +1 @@
-"http://example.org/avatar.jpg"
+"http://://example.org:80/avatar.jpg"
```

Nope, seems like this bug has aways been around. 😞 

Okay, we now know when - it's always been broken. The why that obviously follows on is that no-one wrote a test for it.

So what about fixing it? Well, since `https: true` works then it seems a bit unfortunate that to fix the case where people haven't set anything we need to get the people who did to change their code. If we search elsewhere in the Rails code base we find this line in [action_dispatch/testing/integration.rb][1]

``` ruby
"rack.url_scheme" => https? ? "https" : "http"
```

Can we use a similar technique to fix this problem? Turns out if we change the `normalize_keys` method to this:

``` ruby
      def normalize_keys(env)
        new_env = {}
        env.each_pair { |k,v| new_env[rack_key_for(k)] = rack_value_for(k, v) }
        new_env['rack.url_scheme'] = new_env['HTTPS'] == 'on' ? 'https' : 'http'
        new_env
      end
```

and re-run our test then we get:

```
.

Finished in 0.130983s, 7.6346 runs/s, 7.6346 assertions/s.

1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
```

🎉 - the bug is fixed without anyone having to change their code.

So @tzabaman can you update your PR with this fix and remove the deprecation? Thanks 👍 

[1]: https://github.com/rails/rails/blob/416d85b65e4f49f7e86f24c56bc1ec7441e90f2c/actionpack/lib/action_dispatch/testing/integration.rb#L233

**Comment by pixeltrix** [comment]
@tzabaman I think something went wrong with your rebase - it's showing 86 commits.

**Comment by tzabaman** [comment]
@pixeltrix Maybe that happened because I rebased my branch with rails master... Do you propose to open a new PR and drop this one?

**Comment by pixeltrix** [comment]
@tzabaman try squashing all your commits into a single commit and then rebase that one against master

**Comment by tzabaman** [comment]
@pixeltrix Thank you very much for your precious help! The methodology with which you approached and finally solved the issue will be very useful for future similar situations! :+1:

**Comment by pixeltrix** [comment]
@tzabaman thanks! 👍

**Comment by pixeltrix** [comment]
Backported to 5-0-stable in f69edbea34c03aad7d3797cea8a4101dc3040068

**Comment by pixeltrix** [comment]
@tzabaman for future reference we normally add a CHANGELOG entry, thanks again.

**Comment by tzabaman** [comment]
@pixeltrix Thank you, I will have it in mind next time!

**Pull request merged**